src/Controllers/cli_controller.py

```diff
@@ -26,11 +30,161 @@ def parse_list_cmd(self, d:UML_Diagram, tokens:list[str]):
+        return res
+
+    def get_add_relation(self) -> dict[str, dict[str, str]]:
+        """
+        Suggests class, class, relationship type 
+
+        Returns:
+        dict[str, dict[str, str]]: A dictionary where keys are class names (str)
+                                                 and values are dictionaries of class names (str) and
+                                                 corresponding relationship types (str).
+        """
+        rel_types = dict([
+            ("Aggregation", None),
+            ("Composition", None),
+            ("Realization", None),
+            ("Inheritance", None)
+        ])
+        class_reltype = dict((key, rel_types) for key in self.get_classes())
+        return dict((key, class_reltype) for key in self.get_classes())
+
+    def get_classes(self) -> dict[str, None]:
+        """
+        Get a dictionary of class names.
+
+        Returns:
+            dict[str, None]: An dictionary where keys are class names (str) and values are None.
+        """
+        return dict((c.get_name(), None) for c in self._diagram.get_all_classes())
+
+    def get_methods(self) -> dict[str, set[str]]:
+        """
+        Get a dictionary of methods for each class.
+
+        Returns:
+            dict[str, set[str]]: A dictionary where keys are class names (str) and
+                                        values are sets containing names of methods (str) for each class.
+        """
+        res = dict()
+        for uml_class in self._diagram.get_all_classes():
+            methods = set()
+            for method in uml_class.get_methods():
+                methods.add(method.get_name())
+            if methods:
+                res[uml_class.get_name()] = methods
+        return res
+
+    def get_fields(self) -> dict[str, set[str]]:
+        """
+        Get a dictionary of fields for each class.
+
+        Returns:
+            dict[str, set[str]]: A dictionary where keys are class names (str) and
+                                        values are sets containing names of fields (str) for each class.
+        """
+        res = dict()
+        for uml_class in self._diagram.get_all_classes():
+            fields = set()
+            for field in uml_class.get_fields():
+                fields.add(field.get_name())
+            if fields:
+                res[uml_class.get_name()] = fields
+        return res
+
+    def get_params(self) -> dict[str, dict[str, set[str]]]:
+        """
+        Retrieve parameters of methods in each class.
+
+        Returns:
+            dict[str, dict[str, set[str]]]: A dictionary where keys are class names and values
+            are dictionaries containing method names as keys and sets of parameter names as values.
+        """
+        res = dict()
+        for uml_class in self._diagram.get_all_classes():
+            nested_dict = None
+            for method in uml_class.get_methods():
+                params = set()
```

Comment:
instead of looping through like this, use list's set conversion. (see above comments for syntax)

Reply:
same as above